metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj

```diff
@@ -589,3 +590,79 @@
                  concept-id-rev-ids-map)))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Functions for working with cascading deletes and associations
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn get-tombstone-transaction
+  "Get the transaction for a given tombstone."
+  [db {:keys [concept-id revision-id]} provider-id concept-type]
+  (-> db
+      (concepts/get-concept concept-type
+                           {:provider-id provider-id :small false}
+                           concept-id
+                           revision-id)
+      :transaction-id))
+
+(defn get-tombstone-associations
+  "A convenience function for querying the database for tombstone
+  associations. By default, associations use the 'CMR' provider-id; if this
+  is not what you need, use the 3-arity function and pass the desired
+  provider."
+  ([db tombstone]
+   (get-tombstone-associations db tombstone {}))
+  ([db tombstone params]
+   (get-tombstone-associations db tombstone "CMR" params))
+  ([db tombstone provider-id params]
+   (concepts/find-latest-concepts db {:provider-id provider-id} params)))
+
+(defn only-newer-associations
+  "Filter out any associations older than the tombstone."
+  [db tombstone provider-id concept-type params]
+  (filter #(< (:transaction-id %)
+              (get-tombstone-transaction db tombstone provider-id concept-type))
+          (get-tombstone-associations db tombstone params)))
+
+(defn publish-delete-associations
+  "Publish events for concept associatiosn deletes. The in-memory database uses
+  this function without the provider-id to publish events only; it doesn't save
+  the concepts as well (the external database does)."
+  ([db tombstones]
+    (publish-delete-associations db tombstones nil))
+  ([db provider-id tombstones]
+   (doseq [tombstone tombstones]
+     (when provider-id
+       (concepts/save-concept db
+                              {:provider-id provider-id :small false}
+                              tombstone))
+     ;; publish association delete event
+     (ingest-events/publish-event
+       (:context db)
+       (ingest-events/concept-delete-event tombstone)))))
+
+(defn update-tombstone-associations
+  "A utility function for use when performing cascading deletes on concept
+  associations."
+  [associations]
+  (map #(-> %
+            (assoc :deleted true :metadata "")
+            (update :revision-id inc))
+        associations))
+
+(defn cascade-delete-concept-associations
+  "Save tombstones for all the associations of the given concept tombstone."
+  [db provider tombstone params]
+  (let [concept-type (:concept-type tombstone)
+        provider-id (or (:provider-id provider) "CMR")]
+    (->> params
+         (only-newer-associations db tombstone provider-id concept-type)
+         (update-tombstone-associations)
+         (publish-delete-associations db provider-id))))
+
+(defmulti cascade-delete-associations
```

Comment:
No worries! I shared with you on a call, but for others that read this PR -- my comment above "I think the diff for the suggested refactor is actually larger that the original change ;-)" was truly meant in good humour and not snarky at all :-)

In fact, before that refactoring I had planned on adding chunks of this code to the CMR dev pain points on the wiki page ... and now the issues I identified are no longer present -- thanks to @ygliuvt! :-D